File: 0739-daily-temperatures/0739-daily-temperatures.cpp
// TC- O(n)
// SC- O(n)

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        // monotonically decreasing array
        vector<int> res(temperatures.size(), 0);
        stack<int> st;

        for(int t=0; t<temperatures.size(); t++){
            while(!st.empty() and temperatures[st.top()]<temperatures[t]){
                res[st.top()] = t-st.top();
                st.pop();
            }
            st.push(t);
        }
        return res;
    }
};

File: 0739-daily-temperatures/0739-daily-temperatures.py
# TC- O(n)
# SC- O(n)

class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack = []
        ans = [0]*len(temperatures)
        for i, temp in enumerate(temperatures):
            while(stack and temperatures[stack[-1]]<temp):
                j = stack.pop()
                ans[j] = i-j
            stack.append(i)
        return ans

            